File: ActualCode/NewSwerveTest-DO_NOT_LOOK-Imported/src/main/cpp/DriveManipulation.cpp
#include <numbers>
#include <frc/smartdashboard/SmartDashboard.h>
#include <frc/XboxController.h>

#include "DriveManipulation.h"


DriveManipulation::DriveManipulation(frc::XboxController* getController) {
    controller = getController;
}

void DriveManipulation::setNewCenterState() {
    x = controller->GetLeftX();
    y = -controller->GetLeftY();
    rotation = controller->GetRightX();

    // V_p
    double positionVector[2] = {x, y};
    // v_s
    double newRotationVector[4][2];
    double constantRotationVector[4][2];

    // we need to get the x and y components of the rotation vector
    for (int i = 0; i < 4; i++) {
        newRotationVector[i][0] = std::cos(constantRotationAngle[i]);
        newRotationVector[i][1] = std::sin(constantRotationAngle[i]);
    }

    // we then need to scale the rotation vector by the rotation value
    for (int i = 0; i < 4; i++) {
        newRotationVector[i][0] *= rotation;
        newRotationVector[i][1] *= rotation;
    }

    // now that we have the two scaled and set vector arrays, we can add them together

    // V_f
    double finalVector[4][2];
    for (int i = 0; i < 4; i++) {
        finalVector[i][0] = positionVector[0] + newRotationVector[i][0];
        finalVector[i][1] = positionVector[1] + newRotationVector[i][1];
    }

    // now we can get the angle from the position vector
    for (int i = 0; i < 4; i++) {
        // we dont use atan2 because we want to keep the angle between 0 and 2pi
        if (finalVector[i][1] > 0) {
            swerveModuleAngles[i] = std::acos(finalVector[i][0] / std::sqrt(finalVector[i][0] * finalVector[i][0] + finalVector[i][1] * finalVector[i][1]));
        } else if (finalVector[i][1] < 0) {
            swerveModuleAngles[i] = std::numbers::pi * 2 - std::acos(finalVector[i][0] / std::sqrt(finalVector[i][0] * finalVector[i][0] + finalVector[i][1] * finalVector[i][1]));
        } 
    }

    // dont forget to add the offset
    for (int i = 0; i < 4; i++) {
        swerveModuleAngles[i] += swerveAngleOffset[i] + (std::numbers::pi /2);
    }

}

void DriveManipulation::runToState() {
    frontLeft.setDesiredAngle(swerveModuleAngles[0]);
    frontRight.setDesiredAngle(swerveModuleAngles[1]);
    backLeft.setDesiredAngle(swerveModuleAngles[2]);
    backRight.setDesiredAngle(swerveModuleAngles[3]);

    frontLeft.runToState();
    frontRight.runToState();
    backLeft.runToState();
    backRight.runToState();
}

double DriveManipulation::getSwerveModuleAngle(int module) {
    switch (module) {
        case 0:
            return frontLeft.getCurrentAngle();
            break;
        case 1:
            return frontRight.getCurrentAngle();
            break;
        case 2:
            return backLeft.getCurrentAngle();
            break;
        case 3:
            return backRight.getCurrentAngle();
            break;
        default:
            return 0;
            break;
    }
}

// double DriveManipulation::averageAngles(double angle1, double angle2, double weight1, double weight2) {
//     double angle = 0;
//     double viableAngle1[3] = {
//         angle1,
//         angle1 + std::numbers::pi * 2,
//         angle1 - std::numbers::pi * 2
//     };
//     double viableAngle2[3] = {
//         angle2,
//         angle2 + std::numbers::pi * 2,
//         angle2 - std::numbers::pi * 2
//     };
//     double differences[9] = {
//         viableAngle1[0] - viableAngle2[0],
//         viableAngle1[0] - viableAngle2[1],
//         viableAngle1[0] - viableAngle2[2],
//         viableAngle1[1] - viableAngle2[0],
//         viableAngle1[1] - viableAngle2[1],
//         viableAngle1[1] - viableAngle2[2],
//         viableAngle1[2] - viableAngle2[0],
//         viableAngle1[2] - viableAngle2[1],
//         viableAngle1[2] - viableAngle2[2]
//     };
//     double smallestDifference[2];
//     // ablsolute the entire differences array
//     for (int i = 0; i < 9; i++) {
//         differences[i] = abs(differences[i]);
//     }
//     // find the index that holds the smallest difference
//     int mIndex = std::min_element(differences, differences + 9) - differences;

//     // use the index to use the smallest difference
//     switch (mIndex)
//     {
//     case 0:
//         smallestDifference[0] = viableAngle1[0];
//         smallestDifference[1] = viableAngle2[0];
//         break;
//     case 1:
//         smallestDifference[0] = viableAngle1[0];
//         smallestDifference[1] = viableAngle2[1];
//         break;
//     case 2:
//         smallestDifference[0] = viableAngle1[0];
//         smallestDifference[1] = viableAngle2[2];
//         break;
//     case 3:
//         smallestDifference[0] = viableAngle1[1];
//         smallestDifference[1] = viableAngle2[0];
//         break;
//     case 4:
//         smallestDifference[0] = viableAngle1[1];
//         smallestDifference[1] = viableAngle2[1];
//         break;
//     case 5:
//         smallestDifference[0] = viableAngle1[1];
//         smallestDifference[1] = viableAngle2[2];
//         break;
//     case 6:
//         smallestDifference[0] = viableAngle1[2];
//         smallestDifference[1] = viableAngle2[0];
//         break;
//     case 7:
//         smallestDifference[0] = viableAngle1[2];
//         smallestDifference[1] = viableAngle2[1];
//         break;
//     case 8:
//         smallestDifference[0] = viableAngle1[2];
//         smallestDifference[1] = viableAngle2[2];
//         break;
//     default:
//         break;
//     }

//     // now we can average the angles
//     angle = (smallestDifference[0] * weight1 + smallestDifference[1] * weight2) / (weight1 + weight2);

    
//     return angle;
// }

File: ActualCode/SwerveCode/src/main/java/frc/robot/DriveManipulation.java
package frc.robot;

import com.revrobotics.CANSparkFlex;
import edu.wpi.first.wpilibj.XboxController;
import frc.robot.TurnMotor;

public class DriveManipulation {
    private double[] swerveModuleAngles = {0, 0, 0, 0}; // in radians
    private double[] swerveModuleSpeeds = {0, 0, 0, 0}; // in meters per second
    // Angle for the translation of the robot/ module derived from position
    private double angleFromPosition;
    
    private double x;
    private double y;
    private double rotation;


    // These are constants that needn't be changed by any code

    // offset for where motors are in radians
    private double[] swerveAngleOffset = {4.2107776536,5.595962,6.062292,0.503146};

    /**
     * If the robot rotates, there are different angles that the robot uses to rotate.
     * In the case that the robot is to rotate, we need to use the vector that aligns itself with the robot's rotation
     * These are the default radian measure of the robot
     * @category Constant Expression */
    private double[] constantRotationAngle = {
        - Math.PI * 3 /4,
        - Math.PI / 4,
        - Math.PI * 5 / 4 ,
        - Math.PI * 7 / 4
    };

    private XboxController controller;
    private TurnMotor frontLeft  = new TurnMotor( 9, 16);
    private TurnMotor frontRight = new TurnMotor(11, 17);
    private TurnMotor backLeft   = new TurnMotor(13, 18);
    private TurnMotor backRight  = new TurnMotor(15, 19);

    private CANSparkFlex frontLeftDrive  = new CANSparkFlex( 8, com.revrobotics.CANSparkLowLevel.MotorType.kBrushless);
    private CANSparkFlex frontRightDrive = new CANSparkFlex(10, com.revrobotics.CANSparkLowLevel.MotorType.kBrushless);
    private CANSparkFlex backLeftDrive   = new CANSparkFlex(12, com.revrobotics.CANSparkLowLevel.MotorType.kBrushless);
    private CANSparkFlex backRightDrive  = new CANSparkFlex(14, com.revrobotics.CANSparkLowLevel.MotorType.kBrushless);

    DriveManipulation(XboxController getController) {
        controller = getController;
    }
    void setNewCenterState() {
        x        =  controller.getLeftX();
        y        = -controller.getLeftY();
        rotation =  controller.getRightX();

        if ((x < .1 && x > -.1) && (y < .1 && y > -.1) && (rotation < .1 && rotation > -.1)) {
            x = 0;
            y = 0;
            rotation = 0;
            for (int i = 0; i < 4; i++) {
                swerveModuleSpeeds[i] = 0;
            }
        } else {
            // V_p
            double[] positionVector = {x, y};
            // v_s
            double[][] newRotationVector = new double[4][2];
            double[][] constantRotationVector = new double[4][2];

            // we need to get the x and y components of the rotation vector
            for (int i = 0; i < 4; i++) {
                newRotationVector[i][0] = Math.cos(constantRotationAngle[i]);
                newRotationVector[i][1] = Math.sin(constantRotationAngle[i]);
            }

            // we then need to scale the rotation vector by the rotation value
            for (int i = 0; i < 4; i++) {
                newRotationVector[i][0] *= rotation;
                newRotationVector[i][1] *= rotation;
            }

            // now that we have the two scaled and set vector arrays, we can add them together

            // V_f
            double[][] finalVector = new double[4][2];
            for (int i = 0; i < 4; i++) {
                finalVector[i][0] = positionVector[0] + newRotationVector[i][0];
                finalVector[i][1] = positionVector[1] + newRotationVector[i][1];
            }

            // now we can get the angle from the position vector
            // for (int i = 0; i < 4; i++) {
            //     // we dont use atan2 because we want to keep the angle between 0 and 2pi
            //     if (finalVector[i][1] > 0) {
            //         swerveModuleAngles[i] = Math.acos(finalVector[i][0] / Math.sqrt(finalVector[i][0] * finalVector[i][0] + finalVector[i][1] * finalVector[i][1]));
            //     } else if (finalVector[i][1] < 0) {
            //         swerveModuleAngles[i] = Math.PI * 2 - Math.acos(finalVector[i][0] / Math.sqrt(finalVector[i][0] * finalVector[i][0] + finalVector[i][1] * finalVector[i][1]));
            //     } 
            // }

            // calculate the hypotenuse
            double hypotenuse = Math.sqrt(finalVector[0][0] * finalVector[0][0] + finalVector[0][1] * finalVector[0][1]);
            // calculate the angle using the previously commented out for loop, use hypotenuse
            for (int i = 0; i < 4; i++) {
                if (finalVector[i][1] > 0) {
                    swerveModuleAngles[i] = Math.acos(finalVector[i][0] / hypotenuse);
                } else if (finalVector[i][1] < 0) {
                    swerveModuleAngles[i] = Math.PI * 2 - Math.acos(finalVector[i][0] / hypotenuse);
                } 
            }

            // dont forget to add the offset
            for (int i = 0; i < 4; i++) {
                swerveModuleAngles[i] += swerveAngleOffset[i] + (Math.PI /2);
            }

            // now that al the angles are set, we can set the speeds
            for (int i = 0; i < 4; i++) {
                swerveModuleSpeeds[i] = Math.sqrt(finalVector[i][0] * finalVector[i][0] + finalVector[i][1] * finalVector[i][1]);
            }}
    }
    void runToState() {
        frontLeft.setDesiredAngle(swerveModuleAngles[0]);
        frontRight.setDesiredAngle(swerveModuleAngles[1]);
        backLeft.setDesiredAngle(swerveModuleAngles[2]);
        backRight.setDesiredAngle(swerveModuleAngles[3]);
    
        frontLeft.runToState();
        frontRight.runToState();
        backLeft.runToState();
        backRight.runToState();
    
        // now we can set the speeds
        frontLeftDrive.set(swerveModuleSpeeds[0]);
        frontRightDrive.set(swerveModuleSpeeds[1]);
        backLeftDrive.set(swerveModuleSpeeds[2]);
        backRightDrive.set(swerveModuleSpeeds[3]);
    }

}
